### PR TITLE
use v0.44: no core dumps, OSX as well

### PR DESCRIPTION
**Pull request by dneise**
This should solve #11  and #12

### CODE REVIEW

**Comment by dneise** [comment]
Okay so the test for OSX confirms, this version does not work on OSX. (As expected), no I put the *darwin.so files and the dylib files from the tarballs into this repo and try again.

**Comment by calispac** [comment]
What is going on here? 
https://travis-ci.org/cta-sst-1m/protozfitsreader/jobs/338868451#L28-L33

**Comment by dneise** [comment]
> What is going on here?
https://travis-ci.org/cta-sst-1m/protozfitsreader/jobs/338868451#L28-L33

I think at that point OSX tries to install Python 3.5 and fails. I have really no experience with OSX, so ... I thought I let travis simply test the stuff. 
I admit, I expected it to fail and did not even bother to look why it failed. But it looks like it failed even before trying to install the protozfitsreader ... that is interesting.

**Comment by dneise** [comment]
Okay so now it nicely fails on OSX and it still works on linux, nice.
I actually added the OSX binaries already, but the system (as usual) does not find the binaries. 

An issue, which I tried to solve by modifying the RPATH inside the .so files manually.

On OSC the tool to modify the RPATH inside the .so and .dylib files, so they do find each other is called 
`install_name_tool` apparently: https://stackoverflow.com/questions/33991581/install-name-tool-to-update-a-executable-to-search-for-dylib-in-mac-os-x

Since I do not have a Mac and do not know anybody who has a Mac, I cannot modify the files accordingly. I think.

**Comment by dneise** [comment]
I think I could at least try to set the `DYLD_LIBRARY_PATH` on OSX and see if it works then.

**Comment by calispac** [comment]
> Since I do not have a Mac and do not know anybody who has a Mac, I cannot modify the files accordingly. I think.

Sorry me neither.

**Comment by dneise** [comment]
Okay .. so at this point in time. I was able to have OSX stuff running (fixes #12), abeit with the end using having to `export DYLD_LIBRARY_PATH=/path/to/anaconda/env/lib/python3.x/site-packages` or so ... c.f. travis.yml.

And I have added a test, that shows a SegFault when iterating too far:
e.g. https://travis-ci.org/cta-sst-1m/protozfitsreader/jobs/339074549#L667

This test still fails in the linux tests at the moment, since they are still using the v0.43 rawzfitsreader.so files (even though stupid me already updated the setup.py to 0.44, so setup.py is lying at the moment).

In the next push I update the linux rawzfitsreaders to 0.44 as well. If then all tests pass, I am (quite) happy.

**Comment by dneise** [comment]
Okay This fixes now #11  and #12 (partly)

File: README.md [hunk 1]
```diff
@@ -2,8 +2,33 @@
 
 ## installation:
```
Comment (calispac) [inline README.md]: maybe here a sentece explaining that you should have anaconda installed and either create an env or use an existing one?

File: README.md [hunk 2]
```diff
@@ -2,8 +2,33 @@
...
 
+
+### Linux (with anaconda)
+
+    conda install numpy protobuf libgcc
```
Comment (calispac) [inline README.md]: `libgcc` is not required by travis. so should we remove it ? also might be better to create a `requirements.txt` file for pip no?
`pytest` is missing

File: setup.py [hunk 3]
```diff
@@ -3,12 +3,15 @@
...
     author_email="etienne.lyard@unige.ch",
     package_data={
-        'protozfitsreader': ['*.so*'],
+        'protozfitsreader': [
+            '*.so*',
+            '*.dylib'
+        ],
         '': ['tests/resources/*'],
     },
     install_requires=['numpy', 'protobuf'],
```
Comment (calispac) [inline setup.py]: update requires. maybe read it from requirements.txt?

**Comment by dneise** [comment]
No that file was apparently accidentally added to the repo previously. 
It is a build artifact and should not be added to the repo. 

Therefire it was removed in this PR and a .gitignorefile was added, which might help avoiding accidentally adding build artifacts in the future.

File: README.md (hunk 2 above)
Comment (dneise) [reply README.md]: Well spotted this `libgcc` stuff ... I removed it.

File: setup.py (hunk 3 above)
Comment (dneise) [reply setup.py]: regarding install_requires and requirements.txt I just read:
https://pip.pypa.io/en/stable/user_guide/#requirements-files
and 
https://caremad.io/posts/2013/07/setup-vs-requirement/

again. I fill follow what Donald Stufft wrote there.

File: README.md (hunk 2 above)
Comment (dneise) [reply README.md]: pytest is not an install requirement ..  I'll add it to setup.py as recommended by the pytest documentation

File: README.md (hunk 1 above)
Comment (dneise) [reply README.md]: I added some text here. Have a look please.

**Comment by dneise** [comment]
Okay, so ...  I should say something about conda and pip and setup.py and requirements.txt and environment.yml here.

The simplest solution here, in my opinion is the best. The simplest for a user is:
`pip install protozfits`

No  `git clone, conda env, pip install requirements.txt, conda install` 

So in the travis test I install now (as a test) nothing from conda ... but let pip install everything.

pip install is much slower for `numpy` and I have never tried to `pip install numpy`. For protobuf the same is true, I have never tried that.

However. Generally I think this package should be pip installable. But we just add a *recommendation* to `conda install numpy protobuf`before doing pip install. Since it is just faster.

Also please note, in general people using this library will for sure have anaconda ... unless there is something better than anaconda in 2025. 

Have a look at the list of packages in anaconda:
https://docs.anaconda.com/anaconda/packages/py3.6_linux-64

You'll see than protobuf and numpy already come with anaconda ... so this package in principle "has no dependencies" (apart from the huge dependency called "Anaconda"). So a "normal vanilla physicist" who installed Anaconda during her first year in university on her laptop, should not have a problem installing this. That is my goal.

In this sense, trying to install numpy and protobuf using pip and waiting for hours ... is not needed...

**Comment by Etienne12345** [comment]
Getting rid of pip entirely, and putting everything in conda remains on my todo list.... 

It sounds like a bad idea to remove the dependencies from the pip package, because the package does depend on these: if someone doesn't use Anaconda, then these deps should be resolved. 

There is a clear incompatibility here between deps / no deps: I'll try to raise the priority of the condaification of the package :)

**Comment by dneise** [comment]
> Getting rid of pip entirely, and putting everything in conda remains on my todo list....

I do not understand this. I know about no package, which is not pip installable, but conda installable.

**Comment by dneise** [comment]
> It sounds like a bad idea to remove the dependencies from the pip package, because the package does depend on these: if someone doesn't use Anaconda, then these deps should be resolved.

Yes ... as I said: This package must be simply pip installable also without conda.

**Comment by Etienne12345** [comment]
Oh ok, possibly: I am bad at conda packages. My point was that if the package is distributed via conda, then problem solved: no dependency needed. As long as we keep it separate, then conflict.

**Comment by dneise** [comment]
> Oh ok, possibly: I am bad at conda packages. My point was that if the package is distributed via conda, then problem solved: no dependency needed. As long as we keep it separate, then conflict.

I do not understand, you said above:  

> if someone doesn't use Anaconda, then these deps should be resolved.

If someone does not use Anaconda then distributing this via conda does not help at all, does it? 
I am utterly confused which direction your arguments are pointing here:
 * One moment your arguments sound like the arguments of pip-purists: "everything must work via pip, we cannot expect everybody to have anaconda" 
 * The next they sound like an Anaconda purist: "when we distribute via conda, we got no problems"

I agree technically with both points: A package must be pip installable, since we simply cannot know if people still use conda in a year from now. Python and pip seem much longer lived and much more stable.

But: installing numpy with pip, takes ages and **nobody** does that. So telling people:
"just use conda to install the deps" is a smart move, in my opinion.

People can also use their package manager to install numpy and protobuf, but this is nothing we should tell them how to do. It's up to them.

For example on ubuntu one can do:

    sudo apt-get install python-numpy

To get some nicely outdated version of numpy ... it is much better than trying to `pip install numpy` but still less good than "conda install numpy".

**Comment by Etienne12345** [comment]
Maybe I missed your first point entirely: I thought that you wanted to remove the deps from the pip package. These should stay IMO, as indeed this package depends on them. How they are resolved is an entirely different story.

**Comment by dneise** [comment]
Ahaaa ... while I was only talking about how to resolve them ... now it's clear why we misunderstood each other.

**Pull request merged**